packages/evm/emulator.go

```diff
@@ -113,28 +137,65 @@ func (e *EVMEmulator) Close() error {
 	return nil
 }
 
-func (e *EVMEmulator) HasPendingBlock() bool {
-	return len(e.pendingBlock.Transactions()) > 0
-}
-
 // Commit imports all the pending transactions as a single block and starts a
 // fresh new state.
 func (e *EVMEmulator) Commit() {
-	if _, err := e.blockchain.InsertChain([]*types.Block{e.pendingBlock}); err != nil {
-		panic(err) // This cannot happen unless the simulator is wrong, fail in that case
+	if len(e.pending.txs) == 0 {
+		return
+	}
+	if _, err := e.blockchain.InsertChain([]*types.Block{e.finalizeBlock()}); err != nil {
+		panic(err)
 	}
-	e.Rollback()
+	e.Rollback(e.pending.header.Time + timeDelta)
 }
 
-func (e *EVMEmulator) newBlock(f func(int, *core.BlockGen)) (*types.Block, []*types.Receipt) {
-	blocks, receipts := core.GenerateChain(e.blockchain.Config(), e.blockchain.CurrentBlock(), ethash.NewFaker(), e.database, 1, f)
-	return blocks[0], receipts[0]
+func (e *EVMEmulator) finalizeBlock() *types.Block {
+	block, err := e.engine.FinalizeAndAssemble(
+		e.blockchain,
+		e.pending.header,
+		e.pending.state,
+		e.pending.txs,
+		nil,
+		e.pending.receipts,
+	)
+	if err != nil {
+		panic(err)
+	}
+	root, err := e.pending.state.Commit(e.blockchain.Config().IsEIP158(e.pending.header.Number))
+	if err != nil {
+		panic(err)
+	}
+	if err := e.pending.state.Database().TrieDB().Commit(root, false, nil); err != nil {
+		panic(err)
+	}
+	return block
 }
 
 // Rollback aborts all pending transactions, reverting to the last committed state.
-func (e *EVMEmulator) Rollback() {
-	e.pendingBlock, _ = e.newBlock(func(int, *core.BlockGen) {})
-	e.pendingState, _ = state.New(e.pendingBlock.Root(), e.blockchain.StateCache(), nil)
+func (e *EVMEmulator) Rollback(timestamp uint64) {
+	statedb, err := e.blockchain.State()
+	if err != nil {
+		panic(err)
+	}
+
+	parent := e.blockchain.CurrentBlock()
+	header := &types.Header{
+		Root:       statedb.IntermediateRoot(e.blockchain.Config().IsEIP158(parent.Number())),
```

Comment:
I think we can just pass `true` here
```suggestion
		Root:       statedb.IntermediateRoot(true),
```